clientv3/retry.go

```diff
@@ -54,14 +63,50 @@ func (c *Client) newRetryWrapper() retryRpcFunc {
 	}
 }
 
+func (c *Client) newRdonlyRetryWrapper() retryRpcFunc {
+	return func(rpcCtx context.Context, f rpcFunc) error {
+		for {
+			err := f(rpcCtx)
+			if err == nil {
+				return nil
+			}
+
+			// always stop retry on etcd errors other than invalid auth token
+			eErr := rpctypes.Error(err)
+			if etcdErr, ok := eErr.(rpctypes.EtcdError); ok {
+				if etcdErr.Code() == rpctypes.ErrInvalidAuthToken.(rpctypes.EtcdError).Code() {
+					gterr := c.getToken(rpcCtx)
+					if gterr != nil {
+						return err // return the original error for simplicity
+					}
+					continue
+
+				}
+				return err
+			}
+
+			return err
+		}
+	}
+}
+
 type retryKVClient struct {
 	pb.KVClient
-	retryf retryRpcFunc
+	retryf       retryRpcFunc
+	rdonlyRetryf retryRpcFunc
 }
 
 // RetryKVClient implements a KVClient that uses the client's FailFast retry policy.
 func RetryKVClient(c *Client) pb.KVClient {
```

Comment:
`retryKVClient` is still only for handling failfast/mutable operations. It shouldn't handle Range because the failfast=false policy works fine. I guess it should be renamed to something like `retryFailfasKVClient` or something. Probably have a separate `type retryAuthKvClient struct { pb.KVClient }` to wrap retryKVClient to separate concerns instead of baking auth retry checks into the failfast retry code?